### PR TITLE
Protection against log forging

### LINKED ISSUE #243: Protect against log forging
Issue Description
--
Static scanning of client code revealed a log forging vulnerability in our code relating to logging parameters received from an external source. Looking in BaseRestController it seems FUSE is exposed to this too (as we log the requestId, which is a string).

This obviously isn't a critical issue, but it is good to be aware of these types of issues.

See https://www.baeldung.com/jvm-log-forging for details.

To add more colour, this issue only came up after we delomboked the code before running the scan.

Design
--
[Description of the root cause for the issue and how to resolve it. Should be complete before Review is approved]

Documentation Changes
--
[Description of changes necessary to the Fuse documentation due to the changes in response to this issue. Should be complete before Review is approved]

Test Evidence
--
Before:
image

After:
image

Validation in Develop
-- 
[Proof issue is resolved in Develop. Should be complete before Issue is closed]

### PR DESCRIPTION
**Pull request by Gala-Sortino**
Closes #243

### CODE REVIEW

File: src/main/java/org/galatea/starter/entrypoint/BaseRestController.java [hunk 1]
```diff
@@ -4,21 +4,36 @@
...
 @Log(enterLevel = Level.INFO, exitLevel = Level.INFO)
 public abstract class BaseRestController {
+  static final char ENCODER_SEPERATOR = '_';
+
+  /** 
+   * Encodes specified message, to avoid any log forging.
+   */
+  protected String encode(String message){
+
+    String encodedMessage = message.replace( '\n', ENCODER_SEPERATOR ).replace( '\r', ENCODER_SEPERATOR )
```
Comment (gala-lightfoot) [inline src/main/java/org/galatea/starter/entrypoint/BaseRestController.java]: spelling - SEP**A**RATOR

File: src/main/java/org/galatea/starter/entrypoint/BaseRestController.java [hunk 2]
```diff
@@ -4,21 +4,36 @@
...
+
 
   /**
    * Adds the specified requestId to the context for this request (if not null).
    */
   protected void processRequestId(String requestId) {
     if (requestId != null) {
-      log.info("Request received with id: {}", requestId);
-      Tracer.setExternalRequestId(requestId);
+      log.info("Request received with id: {}", encode(requestId));
```
Comment (gala-dan) [inline src/main/java/org/galatea/starter/entrypoint/BaseRestController.java]: Create an _encodedRequestId_ variable and store the result of calling _encode()_ into it so only call the method once.

File: src/test/java/org/galatea/starter/entrypoint/BaseRestControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+
+import static org.junit.Assert.assertEquals;
+
+@RequiredArgsConstructor
+@Slf4j
+// Use this runner since we want to parameterize certain tests.
+// See runner's javadoc for more usage.
+@RunWith(JUnitParamsRunner.class)
+public class BaseRestControllerTest extends BaseRestController {
```
Comment (gala-dan) [inline src/test/java/org/galatea/starter/entrypoint/BaseRestControllerTest.java:17]: Formatting seems inconsistent in file. Plus make TARGET_ENCODED_MESSAGE private.

File: src/main/java/org/galatea/starter/entrypoint/BaseRestController.java [hunk 4]
```diff
@@ -4,21 +4,37 @@
...
 import org.galatea.starter.utils.Tracer;
-
+import org.owasp.esapi.ESAPI;
 /**
  * Provides base functionality shared by all Fuse REST controllers
  */
 @Slf4j
 @Log(enterLevel = Level.INFO, exitLevel = Level.INFO)
 public abstract class BaseRestController {
+  static final char ENCODER_SEPARATOR = '_';
```
Comment (gala-dan) [inline src/main/java/org/galatea/starter/entrypoint/BaseRestController.java]: Make protected.

File: src/main/java/org/galatea/starter/entrypoint/BaseRestController.java [hunk 5]
```diff
@@ -4,21 +4,37 @@
...
+
+  /** 
+   * Encodes specified message, to avoid any log forging.
+   */
+  protected String encode(String message){
+
+    String encodedMessage = message.replace( '\n', ENCODER_SEPARATOR ).replace( '\r', ENCODER_SEPARATOR )
+      .replace( '\t', ENCODER_SEPARATOR );
+    
+    encodedMessage = ESAPI.encoder().encodeForHTML( encodedMessage );
```
Comment (gala-dan) [inline src/main/java/org/galatea/starter/entrypoint/BaseRestController.java]: Can also squash these two lines into one:

return ESAPI.encoder().encodeForHTML( encodedMessage );

**Comment by gala-dan** [comment]
LGTM, but I'd like someone else to provide an opinion too.

File: src/main/java/org/galatea/starter/entrypoint/BaseRestController.java [hunk 6]
```diff
@@ -4,21 +4,35 @@
...
  */
 @Slf4j
 @Log(enterLevel = Level.INFO, exitLevel = Level.INFO)
 public abstract class BaseRestController {
+  protected static final char ENCODER_SEPARATOR = '_';
+
+  /** 
+   * Encodes specified message, to avoid any log forging.
+   */
+  protected String encode(String message){
```
Comment (gala-dan) [inline src/main/java/org/galatea/starter/entrypoint/BaseRestController.java:21]: Possibly overkill, but we could use the @NonNull annotation on this parameter, given that it should never be called with a null parameter.

Comment (Gala-Sortino) [reply src/main/java/org/galatea/starter/entrypoint/BaseRestController.java:21]: I would argue that it is overkill, provided that we already check for requestID != null in the first place. However, no harm is added by adding this feature and could serve to enforce the method to never be null.

**Comment by GalateaEric** [comment]
Couple comments from the fuse steerco:

1) do these changes now pass the vulnerability security scan as represented in the PR?

2) is the src/main/resources/ESAPI.properties file required for the owasp lib to function?  Looks like it is, but can we trim it down?  Can the parameters in the ESAPI.properties files (in particular the secrets) be passed as cmd line args or env variables?  Certainly we wouldn't want to encourage people to commit a property file with a thing like, "master salt" into Git...

3) is the src/main/resources/validation.properties file required for the owasp lib to function? We're generally of the view that having properties in property files that are then not used anywhere in the code are sub-optimal.  We may want to either remove the properties (or property files) or else provide examples of their use.  

4) are we missing finals for method parameters?  BaseRestController.encode() maybe?

**Review by GalateaEric — CHANGES_REQUESTED** [review]
see comment

**Comment by GalateaEric** [comment]
any further ideas? @Gala-Sortino @gala-dan

**Comment by Gala-Sortino** [comment]
Hey @GalateaEric I have been waiting for @gala-dan to return from vacation to check with him that this fix does indeed pass the scan as you asked in point 1, as I am unsure as to which exact scan was used etc.

Proceeding on to points 2 and 3, as these are configuration files required by ESAPI, it would seem that they are needed to avoid other errors occurring, as ESAPI looks for the information found within these. And lacking in certain information does throw errors. I am currently looking if there is any way around that.

I do agree with the point about passing the masterKey and mastersalt as a cmd line arg so looking at that at the moment, though I am currently unsure on if ESAPI will allow that, as it does seem it is required to be first generated and then put into the ESAPI.properties file.

I agree with point 4 and will make this change.

**Comment by gala-dan** [comment]
For reference the tool used was HPE Security Fortify Audit Workbench. For some reason using ESAPI did not clear the warning in our scan. However, that may be a quirk of the clients setup. Certainly looking at the test evidence in this issue the key vulnerability (the ability for a caller to inject fake log lines) has been guarded against with these changes.

**Comment by GalateaEric** [comment]
Can we get some understanding from Heather (who has some fortify experience) on what was done or needs to be done to clear the error in Fortify.  If the ESAPI import didn't clear it -- what did?

**Comment by GalateaEric** [comment]
closing this PR as I don't think using ESAPI is the way to go.  Despite baeldung's recommendation the ESAPI library doesn't seem to be the way to go for new projects.  In addition to the giant pile of unused properties we would be introducing, there's this article from the main developer back in 2014 indicating the project had issues and wasn't being maintained.

https://www.baeldung.com/jvm-log-forging  -- article originally pointing us to ESAPI
http://off-the-wall-security.blogspot.com/2014/03/esapi-no-longer-owasp-flagship-project.html  -- article indicating ESAPI isn't being maintained

**Pull request closed without merging**